src/base/bittorrent/session.cpp

```diff
@@ -2081,6 +2076,12 @@ LoadTorrentParams Session::initLoadTorrentParams(const AddTorrentParams &addTorr
     else
         loadTorrentParams.category = addTorrentParams.category;
 
+    for (const QString &tag : addTorrentParams.tags)
```

Comment:
@Chocobo1
Lately, I think we're being overly strict about using `auto`. I would use `auto` in cases like this since it doesn't really matter what type is `tag`:
```c++
    for (const auto &tag : addTorrentParams.tags)
    {
        if (hasTag(tag) || addTag(tag))
            loadTorrentParams.tags.insert(tag);
    }
```

Reply:
>I would use auto in cases like this since it doesn't really matter what type is tag:

My opinion is: it is only convenient for writing the `for` loop statement, if there are any usage in the `for` body, it is hard to read out the type quickly (especially when reviewing code from untrusted contributors).

Reply:
>if there are any usage in the for body, it is hard to read out the type quickly (especially when reviewing code from untrusted contributors).

In the example above the usage is type independent. Note, if we'll need to change the tag type to something else (say, a custom Tag type), then the code using `auto` will remain valid without any cares.

Reply:
Another example:
```c++
    for (const Torrent *torrent : asConst(getSelectedTorrents()))
```
We don't see type of container here, so why we have to explicitly declare it in the following modification?
```c++
    const auto selected torrents = getSelectedTorrents();
    for (const Torrent *torrent : selectedTorrents)
```

Reply:
>In the example above the usage is type independent. Note, if we'll need to change the tag type to something else (say, a custom Tag type), then the code using auto will remain valid without any cares.

Yes, but what if the `for` body is changed later (in general) with some more advanced usage of the loop variable? Then the `auto` is really making the type untraceable.

Reply:
> Another example:
> 
> ```c++
>     for (const Torrent *torrent : asConst(getSelectedTorrents()))
> ```
> 
> We don't see type of container here, so why we have to explicitly declare it in the following modification?
> 
> ```c++
>     const auto selectedTorrents = getSelectedTorrents();
>     for (const Torrent *torrent : selectedTorrents)
> ```

I suppose it is fine to use `auto` here, since the container is immutable and unlikely to be modified later.

Reply:
>Yes, but what if the `for` body is changed later (in general) with some more advanced usage of the loop variable? Then the `auto` is really making the type untraceable.

Why we should care about it in advance? It can easily be changed when really needed.

Reply:
The real paradox is that we don't care about this in many cases when we refer to variables declared far from the current point of the code. But for some reason, we require an explicit type when we copy it to a local variable:
```c++
// Torrent *torrent
if (!torrent->category(). isEmpty())
   // ...
```
But why not
```c++
// Torrent *torrent
const auto category = torrent->category();
if (!category. isEmpty())
   // ...
```

Reply:
>Why we should care about it in advance? It can easily be changed when really needed.

If it is easy like you said, then why not get it done at the first time instead of double work?

>The real paradox is that we don't care about this in many cases when we refer to variables declared far from the current point of the code. But for some reason, we require an explicit type when we copy it to a local variable

Good point, many other programming languages do type inference on variables. Perhaps C++ (and this project) should follow this trend too, although I'm concerned it *might* have negative impacts on type readability and letting obscure bugs through.

Reply:
>If it is easy like you said, then why not get it done at the first time instead of double work?

Since it can cause double work in another case:
> if we'll need to change the tag type to something else (say, a custom Tag type), then the code using `auto` will remain valid without any cares.

Consider the example I gave above in two situations:
1. Changing the type
2. Changing the usage in the loop body

In first situation we have nothing to do with this code.
In the second situation we (maybe) need to change loop variable declaration alongside with the body (in the same code location).

If we have loop variable initially declared with explicit type we only save us from small additional change in the second situation. But it cases changes in two (or more) code places in first situation.

Reply:
OK, so how do we decide we need explicit type declaration or not in the case of loop body?

Reply:
>many other programming languages support type inference on variables, perhaps C++ (and this project) should follow this trend too.

👎 
It's bad idea generally. It can be useful only when we have several different representation of the same thing near each other.

I think at least higher level logic should be less dependent from exact types. The discussed code is good example of such a logic.

Reply:
>It's bad idea generally. 

Really? To be specific, I was thinking of typescript and rust in this context.

Reply:
>Really? To be specific, I was thinking of typescript and rust in this context.

Perhaps I didn't understand you. I'm not familiar with these languages. Can you explain what do you mean?
I thought something like `int iFileIndex; QString strTag;`.

Reply:
>OK, so how do we decide we need explicit type declaration or not in the case of the example? Only when it invokes the type/class methods?

I think we can start with this basis.

I even think `auto` could be allowed when only some logically abstract methods are invoked. E.g. when we talk about tag `tag.isEmpty()` or `tag == "some-tag"`can make sense independently of exact type.

We just have to have strong enough reasons to forbid `auto` in some particular case.

Reply:
>Perhaps I didn't understand you. I'm not familiar with these languages. Can you explain what do you mean?
I thought you suggest something like int iFileIndex; QString strTag;.

It is not about inferring types from names but inferring types from the code context.
For example: https://doc.rust-lang.org/rust-by-example/std_misc/channels.html and https://doc.rust-lang.org/rust-by-example/fn/hof.html
Notice that it only declare type at required places and leave it out when the compiler is able to figure it out itself.

Reply:
What I can see from these examples is that the data type is explicitly specified only where it matters. But isn't that what we're talking about? We just have some disagreements about where it's important to specify an explicit type and where it's not, don't we?
>Notice that it only declare type at required places and leave it out when the compiler is able to figure it out itself.

If we only stick to the compiler's ability to "figure out" the data type, then we should just use `auto` wherever it compiles successfully. But that's not what we really want, is it?

Reply:
>What I can see from these examples is that the data type is explicitly specified only where it matters. But isn't that what we're talking about? 

Yes, I was talking off-topic and pointing out this is a trend in other languages and many users are fine with it.

>But that's not what we really want, is it?

Agree.

And more related off-topic: In those other languages IMO reviewing code is especially hard without coding environment ready. Taking a real world program for example: https://github.com/uutils/coreutils/blob/master/src/uu/cp/src/cp.rs#L975-L977
What is the type of those 3 variables by just viewing on github? My guess is:
1. Not sure, std::path or something similar
2. Most likely a bool
3. Not sure, std::path or something similar

My point is: It took me non-trivial amount of mental power to guess the types and I'm still not sure I get them right. IMO on this case, omitting types is only in favor of the code author of saving minuscule typing efforts and creates a high entry requirement for the reviewers (of having proper tools for help reviewing).

Reply:
@Chocobo1 

Many users are fine with it because nowadays, for popular languages, it is trivial to get a proper environment set up thanks to LSP-based tooling.

If you configure with CMake with the `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON`, CMake will generate a compilation database (https://clang.llvm.org/docs/JSONCompilationDatabase.html) that any LSP server (your options for C++ are currently https://clangd.llvm.org/ and https://github.com/MaskRay/ccls) can use to provide you with the required "coding environment" (language-intelligence features) in any editor that has LSP capabilities (there are LSP client extensions available for popular editors).

Note that the project does not actually need to be built beforehand, CMake just needs to complete the configure/generate step.

In a nutshell:

1 - Pick any LSP server, if you're unsure just use `clangd`.
2 - Pick your favorite text editor (Vim, VSCode, Emacs, etc) and install the appropriate LSP plugin to communicate with the LSP server
3 - Generate the `compile_commands.json` file with CMake as explained above.

`clangd` has official LSP plugins for popular editors and a simple and easy tutorial to get started: https://clangd.llvm.org/installation.

This is much simpler and quick than manually configuring paths, IDE settings, etc, and is universal across editors. In fact, the general outline of this workflow is even universal across languages - connect LSP server to client, feed the server the input it needs (which can be auto-generated with the appropriate tooling for each language, such as in the case of CMake and C++), enjoy.

Once you have your editor and LSP server configured, you can get a properly configured environment with advanced language-intelligence features for any CMake-based project instantaneously (after installing any necessary dependency packages):

```bash
$ git clone https://some.project && cd someproject
$ cmake -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON # ... other cmake configure arguments ...
$ # Boom, done
```

You should not be trying to guess types manually unless you absolutely need to.

---

> omitting types is only in favor of the code author of saving minuscule typing efforts and creates a high entry requirement for the reviewers (of having proper tools for help reviewing).

As I've shown, the entry requirement for reviewers is not high with LSP tooling. As for the benefits: the savings on verbosity are significant, which is important to reduce cognitive overhead when reviewing code. Nobody wants to see Java-style `Dog dog = new Dog()` in their code anymore, that amount excessive information only benefits a not-so-smart compiler, not the readers of the code.

Reply:
>My point is: It took me non-trivial amount of mental power to guess the types and I'm still not sure I get them right. IMO on this case, omitting types is only in favor of the code author of saving minuscule typing efforts and creates a high entry requirement for the reviewers (of having proper tools for help reviewing).

My main idea is "we don't need to know the exact types in many cases". Often we operate with more high-level concepts: path, string, iterator, container. As a rule, they either have stable and well-known interfaces, or a self-explanatory use in the code.
Of course, when it comes to lower-level logic that affects the individual detail of specific types/classes, then seeing a specific type makes a sense.

In any case, I don't insist on my worldview. You are free to ask for an explicit type indication where it seems appropriate.